[config] Support configuring additional_endpoints with env vars

olivielpeau opened the pull request:
### What does this PR do?

Adds support for configuring `additional_endpoints` & `logs_config.additional_endpoints` with env vars.

### Additional Notes

This only applies to the infra & log agent for now.

TODO: extend this to the other agents (apm & process)

olivielpeau committed:
[config] Support configuring additional_endpoints with env vars

prognant committed:
logs_config.additional_endpoints configuration through env var

olivielpeau commented on releasenotes/notes/additional-endpoints-config-from-env-vars-a5db51083cbd310f.yaml:
```diff
@@ -0,0 +1,6 @@
+    Add the ability to configure additional endpoints for
+    both the agent and the log agent with environment
+    variables, namely DD_ADDITIONAL_ENDPOINTS and
+    DD_LOGS_CONFIG_ADDITIONAL_ENDPOINTS.
```
let's remove the release note: technically `additional endpoints` is a hidden feature that we don't want to advertise since it has billing implications. Users should only use it after they've contacted support.

prognant replied:
It makes sense, release note dropped.

olivielpeau reviewed (commented):
Could you add tests on the logs config when additional endpoints are set in the config file directly? (I haven't seen any pre-existing ones)

(note: Github won't let me approve the PR since I opened it)

prognant commented:
Good point as #4843 (https://github.com/DataDog/datadog-agent/pull/4843/files#diff-091ace02b49a46f55f2586cfec89a160R855)  broke the logs_config.additional_endpoints parsing as it does not follow the same structure as other *.additional_endpoints.
> Could you add tests on the logs config when additional endpoints are set in the config file directly? (I haven't seen any pre-existing ones)
> 
> (note: Github won't let me approve the PR since I opened it)

prognant committed:
Add tests for logs_config.additional_endpoints

+ fix regression around trailing slashes removal
that should not be perform againts logs_config.
additional_endpoints because it does not share the
same structure than other additional_endpoints config.

prognant committed:
fix gofmt complaints

hush-hush commented on pkg/logs/config/config.go:
```diff
@@ -123,15 +123,20 @@ func buildTCPEndpoints() (*Endpoints, error) {
+		err = json.Unmarshal([]byte(raw), &additionals)
+	} else {
+		err = coreConfig.Datadog.UnmarshalKey("logs_config.additional_endpoints", &additionals)
+	}
```
Could we make this a helper function and avoid code duplication from line 165-171 ?

prognant replied:
Will do.

hush-hush commented on pkg/config/config.go:
```diff
@@ -826,7 +826,6 @@ func trimTrailingSlashFromURLS(config Config) error {
 	var additionalEndpointSelectors = []string{
 		"additional_endpoints",
 		"apm_config.additional_endpoints",
-		"logs_config.additional_endpoints",
```
No more removing trailing slash ?

prognant replied:
Logs_config.additional_endpoints uses a different structure than other addtional_endpoints (cf. https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/config/endpoints.go#L13-L21),  and this loop was overwriting the config key logs_config.additional_endpoints with something inconsistent).
And it uses host & port fields so no need for trailing slash removal.

olivielpeau reviewed (commented):
LGTM, nice find about the (not yet shipped) regression 👍 

I'll let @hush-hush approve this once your last changes are done. Let's ask for a review from Logs-Intake as well for good measure

prognant committed:
Address review

gaetan-deputier reviewed (approved):
lgtm!
